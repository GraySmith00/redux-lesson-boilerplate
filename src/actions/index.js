export const addTodo = (text, id) => ({
  type: 'ADD_TODO',
  text,
  id
});

export const toggleTodo = (id) => ({
  type: 'TOGGLE_TODO',
  id
})
