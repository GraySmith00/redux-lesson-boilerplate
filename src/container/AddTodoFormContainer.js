import React, { Component } from 'react'
import { connect } from 'react-redux'
import { addTodo } from '../actions'

class AddTodoFormContainer extends Component {
  constructor() {
    super()
    this.state =  {
      title: ''
    }
  }

  handleChange = (event) => {
    const { value } = event.target
    this.setState({ title: value })
  }

  submitIdea = (event) => {
    event.preventDefault()
    this.props.handleSubmit(this.state.title, this.props.todos.length)
  }

  render() {
    return (
      <form onSubmit={this.submitIdea}>
        <input type="text"
          placeholder="add an idea"
          onChange={this.handleChange}
        />
        <input type='submit'/>
      </form>
    )
  }
}

const mapStateToProps = reduxStore => {
  return { todos: reduxStore.todos }
}

const mapDispatchToProps = dispatch => {
  return {
    handleSubmit: (title, id) => {
      dispatch(addTodo(title,id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddTodoFormContainer)
