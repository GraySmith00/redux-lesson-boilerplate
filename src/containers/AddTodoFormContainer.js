import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions';

export class AddTodoForm extends Component {
  constructor() {
    super()
    this.state = {
      text: ''
    };
  };

  submitForm = (event) => {
    event.preventDefault();
    this.props.handleSubmit({...this.state, id: Date.now()});
  };

  removedClicked = (event, id) => {
    event.preventDefault();
    this.props.handleRemove(id)
  }

  printTodos = () => {
    const toMap = this.props.todos || []
    return toMap.map((todo, index) => {
      return (
        <div key={`todo-${index}`}>
          <p>{todo.text}</p>
          <button onClick={(e) => this.removedClicked(e, todo.id)}>Remove</button>
        </div>
      )
    })
  }

  render() {
    return (
      <section>
        <form onSubmit={this.submitForm}>
          <input type='text'
                 className='add-text'
                 value={this.state.text}
                 onChange={(e) => this.setState({text: e.target.value})}
          />
          <input type='submit'
                 className='submit'
          />
        </form>
        <div className="todos">
          {this.printTodos()}
        </div>
      </section>
    )
  };
};

export const mapStateToProps =  store => ({
  todos: store.todos
});

export const mapDispatchToProps = dispatch => ({
  handleSubmit: todo => dispatch(actions.addTodo(todo)),
  handleRemove: id => dispatch(actions.removeTodo(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTodoForm)
