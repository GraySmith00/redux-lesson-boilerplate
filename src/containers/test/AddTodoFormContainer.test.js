import React from 'react'
import { shallow } from 'enzyme'
import { AddTodoForm, mapStateToProps, mapDispatchToProps } from '../AddTodoFormContainer'
import * as actions from '../../actions'

describe('AddTodoFormContainer', () => {
  describe('AddTodoForm', () => {
    it('should match the snapshot', () => {
      const renderedComponent = shallow(<AddTodoForm />)
      expect(renderedComponent).toMatchSnapshot()
    })

    it('should call handleSubmit on submitForm', () => {
      const mockHandleSubmit = jest.fn()
      const mockEvent = { preventDefault: jest.fn() }
      const renderedComponent = shallow(<AddTodoForm 
                                          handleSubmit={mockHandleSubmit}
                                        />)
      renderedComponent.instance().submitForm(mockEvent)
      expect(mockHandleSubmit).toHaveBeenCalled()
    })

    it('should call handleRemove on removedClicked', () => {
      const mockHandleRemove = jest.fn()
      const mockEvent = { preventDefault: jest.fn() }
      const renderedComponent = shallow(<AddTodoForm 
                                          handleRemove={mockHandleRemove}
                                        />)
      renderedComponent.instance().removedClicked(mockEvent, 0)
      expect(mockHandleRemove).toHaveBeenCalledWith(0)
    })

    it('should build html when printTodos is called', () => {
      const mockTodos = [{ text: 'Wrestle a wombat', id: 0 }]
      const renderedComponent = shallow(<AddTodoForm 
                                          todos={mockTodos}
                                        />)
      const expected = [
        <div key="todo-0">
          <p>Wrestle a wombat</p>
          <button onClick={(e) => renderedComponent.instance().removedClicked(e, 0)}>
              Remove
          </button>
        </div>
      ]
      expect(renderedComponent.instance().printTodos()).toEqual(expected)
    })
  })

  describe('mapStateToProps', () => {
    it('should pull todos from the store', () => {
      const mockStore = {
        todos: [],
      }
      const result = mapStateToProps(mockStore)
      expect(result.todos).toEqual(mockStore.todos)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch when handleSubmit is called', () => {
      const mockDispatch = jest.fn()
      const mockTodo = { text: 'Pet a platypus', id: 0 }
      const result = mapDispatchToProps(mockDispatch)
      result.handleSubmit(mockTodo)
      expect(mockDispatch).toHaveBeenCalledWith(actions.addTodo(mockTodo))
    })

    it('should call dispatch when handleRemove is called', () => {
      const mockDispatch = jest.fn()
      const result = mapDispatchToProps(mockDispatch)
      result.handleRemove(0)
      expect(mockDispatch).toHaveBeenCalledWith(actions.removeTodo(0))
    })
  })
})
