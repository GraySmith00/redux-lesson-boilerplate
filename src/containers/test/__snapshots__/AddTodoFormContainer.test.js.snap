exports[`AddTodoFormContainer AddTodoForm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddTodoForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}>
          <input
                    className="add-text"
                    onChange={[Function]}
                    type="text"
                    value="" />
          <input
                    className="submit"
                    type="submit" />
</form>,
        <div
          className="todos">
          
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="add-text"
              onChange={[Function]}
              type="text"
              value="" />,
            <input
              className="submit"
              type="submit" />,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "add-text",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "submit",
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "todos",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}>
            <input
                        className="add-text"
                        onChange={[Function]}
                        type="text"
                        value="" />
            <input
                        className="submit"
                        type="submit" />
</form>,
          <div
            className="todos">
            
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="add-text"
                onChange={[Function]}
                type="text"
                value="" />,
              <input
                className="submit"
                type="submit" />,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "add-text",
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "todos",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
